client/consensus/pow/src/worker.rs

```diff
@@ -0,0 +1,217 @@
+	pub fn submit(&mut self, seal: Seal) -> bool {
+		if let Some(build) = self.build.take() {
+			match self.algorithm.verify(
+				&BlockId::Hash(build.metadata.best_hash),
+				&build.metadata.pre_hash,
+				build.metadata.pre_runtime.as_ref().map(|v| &v[..]),
+				&seal,
+				build.metadata.difficulty,
+			) {
+				Ok(true) => (),
+				Ok(false) => {
+					warn!(
+						target: "pow",
+						"Unable to import mined block: seal is invalid",
+					);
+					return false
+				},
+				Err(err) => {
+					warn!(
+						target: "pow",
+						"Unable to import mined block: {:?}",
+						err,
+					);
+					return false
+				},
+			}
+
+			let seal = DigestItem::Seal(POW_ENGINE_ID, seal);
+			let (header, body) = build.proposal.block.deconstruct();
+
+			let mut import_block = BlockImportParams::new(BlockOrigin::Own, header);
+			import_block.post_digests.push(seal);
+			import_block.body = Some(body);
+			import_block.storage_changes = Some(build.proposal.storage_changes);
+
+			let intermediate = PowIntermediate::<Algorithm::Difficulty> {
+				difficulty: Some(build.metadata.difficulty),
+			};
+
+			import_block.intermediates.insert(
+				Cow::from(INTERMEDIATE_KEY),
+				Box::new(intermediate) as Box<dyn Any>
+			);
+
+			match self.block_import.import_block(import_block, HashMap::default()) {
+				Ok(_) => {
+					info!(
+						target: "pow",
+						"✅ Successfully mined block on top of: {}",
+						build.metadata.best_hash
+					);
+					true
+				},
+				Err(err) => {
+					warn!(
+						target: "pow",
+						"Unable to import mined block: {:?}",
+						err,
+					);
+					false
+				},
+			}
+		} else {
+			warn!(
+				target: "pow",
+				"Unable to import mined block: build does not exist",
+			);
+			false
+		}
+	}
+}
+
+pub struct UntilImportedOrTimeout<Block: BlockT> {
+	import_notifications: ImportNotifications<Block>,
+	timeout: Duration,
+	inner_delay: Option<Delay>,
+}
+
+impl<Block: BlockT> UntilImportedOrTimeout<Block> {
+	pub fn new(
+		import_notifications: ImportNotifications<Block>,
+		timeout: Duration,
+	) -> Self {
+		Self {
+			import_notifications,
+			timeout,
+			inner_delay: None,
+		}
+	}
+}
+
+impl<Block: BlockT> Stream for UntilImportedOrTimeout<Block> {
+	type Item = ();
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<()>> {
+		let mut fire = false;
+
+		loop {
+			match Stream::poll_next(Pin::new(&mut self.import_notifications), cx) {
+				Poll::Pending => break,
+				Poll::Ready(Some(_)) => {
+					fire = true;
+				},
+				Poll::Ready(None) => return Poll::Ready(None),
+			}
+		}
+
+		self.inner_delay = match self.inner_delay.take() {
+			None => {
+				Some(Delay::new(self.timeout))
+			},
+			Some(d) => Some(d),
+		};
+
+		if let Some(ref mut inner_delay) = self.inner_delay {
+			match Future::poll(Pin::new(inner_delay), cx) {
+				Poll::Pending => (),
+				Poll::Ready(()) => {
+					fire = true;
+				},
+			}
+		}
+
+		if fire {
+			self.inner_delay = Some(Delay::new(self.timeout));
```

Comment:
I changed this to set `inner_delay` to `None`. If we poll it again here, it really feels redundant as we just polled the previous inner delay above. The line below will return `Poll::Ready`, which means the `poll_next` function will be woken immediately again and then set `inner_delay` back to `Some` and poll it. Precision shouldn't matter as here the timeout is mostly used as a fallback in case the import notifications are not triggered for a long time.

Reply:
Ahh yeah, good point :facepalm: